Add index join support to Thrift connector

AlekseiS opened the pull request.

AlekseiS commented:
Ready for review.

charlesmccarthyirl commented on presto-docs/src/main/sphinx/include/PrestoThriftService.thrift:
```diff
@@ -209,14 +214,18 @@ struct PrestoThriftSplit {
+}
+
+struct PrestoThriftPageResult {
+  1:  PrestoThriftPage page;
```
I'm concerned about this not being backwards compatible. It would require a staggered deployment for existing clients to try to migrate to the new version. The only way I can think for them doing it without downtime would be that they'd need to spin up a new version of their thrift service, update to the new version of presto pointing to that new service, then shut down the old version. 

I'm curious how others feel.

AlekseiS replied:
I discussed it with Dain and we decided to not introduce the new `PrestoThriftPage` class. I'm going to remove it from the PR.

dain commented on presto-thrift-connector-api/src/main/java/com/facebook/presto/connector/thrift/api/PrestoThriftPageResult.java:
```diff
@@ -120,6 +125,21 @@ public String toString()
+        }
+        List<PrestoThriftBlock> thriftBlocks = new ArrayList<>(numberOfColumns);
+        for (int columnIndex = 0; columnIndex < numberOfColumns; columnIndex++) {
+            thriftBlocks.add(fromRecordSetColumn(recordSet, columnIndex, positions));
```
verify that processing the record set multiple times is not "expensive"

AlekseiS replied:
As far as I see, RecordSet for index join is backed by filtered pages with some transformations applied on top. It should be fine overall.

dain commented on presto-thrift-connector-api/src/test/java/com/facebook/presto/connector/thrift/api/TestReadWrite.java:
```diff
@@ -103,36 +108,66 @@
+        });
+    }
+
+    private void testReadWrite(Random random, int records,
```
no need for a line break (or put the all args on multiple lines)

AlekseiS replied:
Done

dain commented on presto-docs/src/main/sphinx/include/PrestoThriftService.thrift:
```diff
@@ -318,6 +318,21 @@ service PrestoThriftService {
 
   /**
+   * Returns a batch of lookup splits.
+   * This method is called if index join is chosen in order to get splits for a given batch of keys.
```
This sentence is a bit confusing.

AlekseiS replied:
Rewrote the comment.

dain commented on presto-docs/src/main/sphinx/include/PrestoThriftService.thrift:
```diff
@@ -318,6 +318,21 @@ service PrestoThriftService {
+   * @param nextToken token from a previous split batch or {@literal null} if it is the first call
+   * @return a batch of splits
+   */
+  PrestoThriftSplitBatch prestoGetLookupSplits(1:  PrestoThriftSchemaTableName schemaTableName, 2:  list<string> lookupColumnNames, 3:  list<string> outputColumnNames, 4:  PrestoThriftPageResult keys, 5:  PrestoThriftTupleDomain outputConstraint, 6:  i32 maxSplitCount, 7:  PrestoThriftNullableToken nextToken) throws (1: PrestoThriftServiceException ex1);
```
Maybe call this `prestoGetIndexSplits`... I have found that most people are confused by the term `looup` in this context

dain replied:
I'd substitute `lookup` for `index` through this code

AlekseiS replied:
When I discussed it with Martin he used "lookup" as a term with an idea to change the way index join is implemented at some point. I'm fine with either name and can change it to "index" for consistency.

AlekseiS replied:
Renamed a method to `getIndexSplits()`. However, I kept parameter name as `lookupColumnNames` as everywhere in presto code they are referred to a lookup, for example `lookupSchema` in `ConnectorIndexProvider`.

dain replied:
I don't think we need to expose our internal implementation details (e.g., lookup) to this interface, so I would name everything index.

AlekseiS replied:
Renamed swift parameter to `indexColumnNames`. Note, that SPI uses the term `lookupColumns`, so I kept this name in implementation.

dain commented on presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/ThriftConnectorIndex.java:
```diff
@@ -0,0 +1,57 @@
+    private final ThriftIndexHandle indexHandle;
+    private final List<ColumnHandle> lookupColumns;
+    private final List<ColumnHandle> outputColumns;
+    private final long maxBytesPerResponse;
```
consider making this `DataSize`

AlekseiS replied:
Config option is a DataSize. Here it's passed as part of the call which is always in bytes.

dain commented on presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/ThriftConnectorIndex.java:
```diff
@@ -0,0 +1,57 @@
+
+import static java.util.Objects.requireNonNull;
+
+public class ThriftConnectorIndex
```
Add a `toString`?

AlekseiS replied:
I see other implementations of connector index (tpch and Calypso) are not providing toString implementation. Do I still need it?

dain replied:
I would expect it will make debugging easier if you expose the `indexHandle`. Other wise when debugging you have to drill in to see what the object is.  IMO the other ones should have a `toString` also

AlekseiS replied:
I think it shouldn't be logged as part of query plan. In any case, I added `toString` method with `indexHandle`, `lookupColumns` and `outputColumns`.

AlekseiS commented on presto-thrift-connector-api/src/main/java/com/facebook/presto/connector/thrift/api/PrestoThriftHostAddress.java:
```diff
@@ -30,7 +32,9 @@
     @ThriftConstructor
-    public PrestoThriftHostAddress(String host, int port)
+    public PrestoThriftHostAddress(
+            @ThriftField(name = "host") String host,
```
I had to use them because of the bug in paranamer.

AlekseiS replied:
Removed.

AlekseiS commented on presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/ThriftTableMetadata.java:
```diff
@@ -0,0 +1,126 @@
+                .map(ThriftColumnHandle.class::cast)
+                .map(ThriftColumnHandle::getColumnName)
+                .collect(toImmutableSet());
+        return indexableKeys.contains(keyColumns);
```
I intended it to be an exact match, because I was not sure if it's true in general case.

AlekseiS commented on presto-thrift-connector-api/src/main/java/com/facebook/presto/connector/thrift/api/PrestoThriftTableMetadata.java:
```diff
@@ -69,20 +67,14 @@ public String getComment()
+     */
+    @Nullable
+    @ThriftField(value = 4, requiredness = OPTIONAL)
+    public List<Set<String>> getIndexableKeys()
```
I felt like set of sets is a weird data structure, especially in other languages. Java API though, guarantees that all set implementations will give consistent equals and hash code, so it works well.

dain replied:
I think it is important to communicate that we do not expect duplicate indexes are not expected

AlekseiS replied:
Added more details to the comment.

AlekseiS commented on presto-thrift-connector-api/src/main/java/com/facebook/presto/connector/thrift/api/PrestoThriftHostAddress.java:
```diff
@@ -81,4 +85,9 @@ public String toString()
                 .toString();
     }
+
+    public static List<HostAddress> toHostAddressList(List<PrestoThriftHostAddress> hosts)
```
Inlined.

AlekseiS commented on presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/ThriftMetadata.java:
```diff
@@ -149,12 +149,12 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
     }
 
-    private ConnectorTableMetadata getTableMetadata(SchemaTableName schemaTableName)
+    private ThriftTableMetadata getTableMetadataChecked(SchemaTableName schemaTableName)
```
Done

AlekseiS commented on presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/ThriftTableMetadata.java:
```diff
@@ -0,0 +1,126 @@
+    public ThriftTableMetadata(SchemaTableName schemaTableName, List<ColumnMetadata> columns, Optional<String> comment, Set<Set<String>> indexableKeys)
+    {
+        this.schemaTableName = requireNonNull(schemaTableName, "schemaTableName is null");
+        this.columns = requireNonNull(columns, "columns is null");
```
Done

AlekseiS commented on presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/ThriftTableMetadata.java:
```diff
@@ -0,0 +1,126 @@
+        this.schemaTableName = requireNonNull(schemaTableName, "schemaTableName is null");
+        this.columns = requireNonNull(columns, "columns is null");
+        this.comment = requireNonNull(comment, "comment is null");
+        this.indexableKeys = requireNonNull(indexableKeys, "indexableKeys is null");
```
Done

AlekseiS commented on presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/TupleDomainConversion.java:
```diff
@@ -0,0 +1,45 @@
+
+public final class TupleDomainConversion
+{
+    private TupleDomainConversion()
```
Done

AlekseiS replied:
Tried that, but it failed checkstyle check with:
LeftCurly: '{' at column 37 should be on a new line

dain replied:
I think you need a single space between the brackets.

AlekseiS replied:
It turns out the key to making it work was no space between brackets.

AlekseiS commented on presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/TupleDomainConversion.java:
```diff
@@ -0,0 +1,45 @@
+                .collect(toImmutableMap(
+                        entry -> ((ThriftColumnHandle) entry.getKey()).getColumnName(),
+                        entry -> fromDomain(entry.getValue())));
+        return new PrestoThriftTupleDomain(thriftDomains);
```
I didn't quite get the comment. Do you suggest inlining `fromDomain` method or change the structure of `PrestoThriftTupleDomain`?

dain replied:
The type declaration`Map<String, PrestoThriftDomain>`... basically inline the variable and you get:
```java
return new PrestoThriftTupleDomain(tupleDomain.getDomains().get().entrySet().stream()
          .collect(toImmutableMap(
                  entry -> ((ThriftColumnHandle) entry.getKey()).getColumnName(),
                  entry -> fromDomain(entry.getValue()))));
```

AlekseiS replied:
Got it!

AlekseiS commented on presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/integration/TestThriftDistributedQueriesIndexed.java:
```diff
@@ -0,0 +1,42 @@
+        // Both of them mean that client closed the connection without reading all the data and are expected in case of index join.
+        // One of the cases it happens is when an index snapshot builder gets out of memory and gets closed. Then index loader will retry or choose another strategy.
+        // Silencing ALL nifty exceptions to not pollute the log.
+        Logging.initialize().setLevel("com.facebook.nifty.core.NiftyExceptionLogger", OFF);
```
Add the code to restore it after test finishes.

AlekseiS commented on presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/TestThriftIndexPageSource.java:
```diff
@@ -0,0 +1,389 @@
+
+    private static <T> List<T> list(T... values)
+    {
+        return ImmutableList.copyOf(values);
```
Inlined.

AlekseiS commented on presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/TestThriftIndexPageSource.java:
```diff
@@ -0,0 +1,389 @@
+        assertEquals(client.timesClosed(), splits + 1);
+    }
+
+    private static class ThriftServiceTestImpl
```
Renamed to `TestingThriftService` and `TestingThriftServiceProvider`.

AlekseiS commented on presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/ThriftIndexProvider.java:
```diff
@@ -0,0 +1,57 @@
+            List<ColumnHandle> lookupSchema,
+            List<ColumnHandle> outputSchema)
+    {
+        ThriftIndexHandle thriftIndexHandle = (ThriftIndexHandle) indexHandle;
```
Done

AlekseiS commented on presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/TestThriftIndexPageSource.java:
```diff
@@ -0,0 +1,389 @@
+    public void testGetNextPageTwoConcurrentRequests()
+            throws Exception
+    {
+        int splits = 3;
```
They are used only in the context of this method, so I'd rather keep them local. I can make them final though to indicate that they don't change.

AlekseiS commented on presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/ThriftIndexPageSource.java:
```diff
@@ -0,0 +1,392 @@
+    }
+
+    @Override
+    public long getSystemMemoryUsage()
```
Moved to the top.

AlekseiS commented on presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/ThriftIndexPageSource.java:
```diff
@@ -0,0 +1,392 @@
+    {
+        long start = System.nanoTime();
+        ListenableFuture<PrestoThriftPageResult> future = context.getClient().getRows(
+                context.getSplit().getSplitId(), outputColumnNames, maxBytesPerResponse, new PrestoThriftNullableToken(nextToken));
```
Done

AlekseiS commented on presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/ThriftIndexPageSource.java:
```diff
@@ -0,0 +1,392 @@
+                splitsClient = null;
+            }
+        }
+        checkState(haveSplits, "must have splits at this point");
```
It is possible to have no splits. This check verifies that there are splits at this point. If there are no splits then it shouldn't even reach here.

dain replied:
In that case I'd change the message to something like `No splits are running`... I don't like the phrase `at this point` in an error message

AlekseiS replied:
I removed this check when I split the method. Changed the message in another error to remove "at this point".

AlekseiS commented on presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/ThriftIndexPageSource.java:
```diff
@@ -0,0 +1,392 @@
+            return;
+        }
+        future.cancel(true);
+        // wait for the future to be cancelled
```
I checked the javadoc for `cancel()` and it confirms that the future will return `true` for `isDone()` once this method returns. So, there's no need to have wait method.

AlekseiS commented on presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/TestThriftIndexPageSource.java:
```diff
@@ -0,0 +1,389 @@
+        int splits = 3;
+        int lookupRequestsConcurrency = 2;
+        int rowsPerSplit = 1;
+        List<SettableFuture<PrestoThriftPageResult>> futures = initializedList(splits, SettableFuture::create);
```
Nice!

AlekseiS commented on presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/TestThriftIndexPageSource.java:
```diff
@@ -0,0 +1,389 @@
+                MAX_BYTES_PER_RESPONSE,
+                lookupRequestsConcurrency);
+
+        pageSource.getNextPage();
```
Done

AlekseiS commented on presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/TestThriftIndexPageSource.java:
```diff
@@ -0,0 +1,389 @@
+                throws PrestoThriftServiceException
+        {
+            if (keys.getRowCount() == 0) {
+                return immediateFuture(new PrestoThriftSplitBatch(list(), null));
```
I don't think so, but in this test it's used as a simple approach to simulate "no splits" situation.

AlekseiS commented on presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/TestThriftIndexPageSource.java:
```diff
@@ -0,0 +1,389 @@
+
+    private static List<List<Integer>> generateKeys(int beginInclusive, int endExclusive)
+    {
+        ImmutableList.Builder<List<Integer>> result = ImmutableList.builder();
```
Done

AlekseiS commented on presto-thrift-testing-server/src/main/java/com/facebook/presto/connector/thrift/server/ThriftIndexedTpchService.java:
```diff
@@ -0,0 +1,205 @@
+            PrestoThriftNullableToken nextToken)
+            throws PrestoThriftServiceException
+    {
+        checkArgument(NUMBER_OF_INDEX_SPLITS <= maxSplitCount, "maxSplitCount for lookup splits is too low");
```
Technically, this should not happen if the engine is working as planned. In case it happens actual calls for data will fail which I think is good enough.

AlekseiS commented on presto-thrift-testing-server/src/main/java/com/facebook/presto/connector/thrift/server/ThriftIndexedTpchService.java:
```diff
@@ -0,0 +1,205 @@
+                splitInfo.getTableName(),
+                schemaNameToScaleFactor(splitInfo.getSchemaName()),
+                ImmutableSet.copyOf(splitInfo.getLookupColumnNames()))
+                .orElseThrow(() -> new IllegalArgumentException("Index is not present"));
```
Done

AlekseiS commented on presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/ThriftIndexPageSource.java:
```diff
@@ -0,0 +1,392 @@
+        if (finished) {
+            return null;
+        }
+        if (!haveSplits) {
```
Moved out splits logic into a separate method.

AlekseiS commented on presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/ThriftIndexPageSource.java:
```diff
@@ -0,0 +1,392 @@
+        resultContext.close();
+
+        // are there more splits available
+        if (splitIndex < splits.size()) {
```
It will create unnecessary wrapping futures when doing initial setup. However, I extracted inner parts of the method, so it should be easier to read now.

AlekseiS commented:
I addressed the comments. Open questions are:
* list of sets for indexable keys
* toString for ThriftConnectorIndex
* exact match for an index or use a prefix
* inline tupleDomain

AlekseiS commented:
Rebased

dain commented:
* list of sets for indexable keys
If it is legal I would change it.
* toString for ThriftConnectorIndex
I would add, but it is your call. It will help with debugging... only handle needs to be shown
* exact match for an index or use a prefix
No need to implement now.  I'd look into what calypso does
* inline tupleDomain
I would, but it is your call (see comment)
* Phrase `at this point` in an error message
I'd change those... They sound like comments to a developer instead of a message that a user may see.
* Remove `lookup` from IDL
I don't think we need to expose our internal implementation details (e.g., lookup) to Thrift since we may change this one day, so I would name everything index.

AlekseiS commented:
@dain , I addressed remaining comments.

AlekseiS commented:
Rebased

AlekseiS commented:
Will merge after release completes.

AlekseiS commented:
Merged.
